### ISSUE TITLE
Nomad Agent panics during restore state

### ISSUE THREAD

**Issue by dvusboy** [hashicorp/nomad#1141]
A client node agent keeps crashing. I'm not sure how to recover from this state. Attempts at draining the node does not help.

### Nomad version
```
Nomad v0.3.2
```
And this time, this executable is the same one as the [published one](https://releases.hashicorp.com/nomad/0.3.2/nomad_0.3.2_linux_amd64.zip). Previously, `rpmbuild` did a `strip -g` on the executable.

### Operating system and Environment details
```
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
```

### Issue
nomad agent keeps crashing. `systemd` journal provides this log snippet:

```
May 02 11:11:36 <host> systemd[1]: Started Nomad is a distributed, highly available, datacenter-aware scheduler.
May 02 11:11:36 <host> systemd[1]: Starting Nomad is a distributed, highly available, datacenter-aware scheduler...
May 02 11:11:36 <host> nomad[1890]: Loaded configuration from /etc/nomad/config.hcl
May 02 11:11:36 <host> nomad[1890]: ==> Starting Nomad agent...
May 02 11:11:36 <host> nomad[1890]: client: using state directory /var/lib/nomad/client
May 02 11:11:36 <host> nomad[1890]: client: using alloc directory /var/lib/nomad/alloc
May 02 11:11:36 <host> nomad[1890]: fingerprint.cgroups: cgroups are available
May 02 11:11:36 <host> nomad[1890]: client: periodically fingerprinting cgroup at duration 15s
May 02 11:11:36 <host> nomad[1890]: fingerprint.consul: consul agent is available
May 02 11:11:36 <host> nomad[1890]: client: periodically fingerprinting consul at duration 15s
May 02 11:11:38 <host> nomad[1890]: fingerprint.env_aws: Error querying AWS Metadata URL, skipping
May 02 11:11:39 <host> nomad[1890]: fingerprint.env_gce: Could not read value for attribute "machine-type"
May 02 11:11:39 <host> nomad[1890]: fingerprint.env_gce: Error querying GCE Metadata URL, skipping
May 02 11:11:39 <host> nomad[1890]: fingerprint.network: Detected interface docker0  with IP 10.1.13.1 during fingerprinting
May 02 11:11:39 <host> nomad[1890]: fingerprint.network: Unable to parse Speed in output of '/usr/sbin/ethtool docker0'
May 02 11:11:39 <host> nomad[1890]: fingerprint.network: Unable to read link speed from /sys/class/net/docker0/speed
May 02 11:11:39 <host> nomad[1890]: fingerprint.network: Unable to read link speed; setting to default 10000
May 02 11:11:39 <host> nomad[1890]: client: applied fingerprints [arch cgroup consul cpu host memory network nomad storage]
May 02 11:11:39 <host> nomad[1890]: driver.docker: using client connection initialized from environment
May 02 11:11:39 <host> nomad[1890]: driver.exec: exec driver is enabled
May 02 11:11:39 <host> nomad[1890]: client: periodically fingerprinting docker at duration 15s
May 02 11:11:39 <host> nomad[1890]: client: available drivers [docker exec raw_exec]
May 02 11:11:39 <host> nomad[1890]: client: periodically fingerprinting exec at duration 15s
May 02 11:11:39 <host> nomad[1890]: client: port not specified, using default port
May 02 11:11:39 <host> nomad[1890]: client: port not specified, using default port
May 02 11:11:39 <host> nomad[1890]: client: port not specified, using default port
May 02 11:11:39 <host> nomad[1890]: client: setting server address list: [r01u01-a.rnd.vsn.local:4647 r01u02-b.rnd.vsn.local:4647 r01u03-c.rnd.vsn.local:4647]
May 02 11:11:39 <host> nomad[1890]: client: alloc "044d16e9-8727-9cb7-21f3-0be80385b169" in terminal status, waiting for destroy
May 02 11:11:39 <host> nomad[1890]: client: alloc "05247bb7-3f93-0920-a169-3a151a8fe3c9" in terminal status, waiting for destroy
May 02 11:11:39 <host> nomad[1890]: client: alloc "0b7aefa7-7d08-27fd-4ad9-afc12f9660d4" in terminal status, waiting for destroy
May 02 11:11:39 <host> nomad[1890]: client: alloc "1140585c-537e-6482-7d2a-f08e5456b4ad" in terminal status, waiting for destroy
May 02 11:11:39 <host> nomad[1890]: client: alloc "14c124ca-7aac-e91a-58eb-d400ce5b9016" in terminal status, waiting for destroy
May 02 11:11:39 <host> nomad[1890]: client: alloc "19a70e70-e514-9649-7482-bf61711d1eb7" in terminal status, waiting for destroy
May 02 11:11:39 <host> nomad[1890]: panic: runtime error: invalid memory address or nil pointer dereference
May 02 11:11:39 <host> nomad[1890]: [signal 0xb code=0x1 addr=0x50 pc=0x534e1b]
May 02 11:11:39 <host> nomad[1890]: goroutine 146 [running]:
May 02 11:11:39 <host> nomad[1890]: panic(0x1124d20, 0xc82000e070)
May 02 11:11:39 <host> nomad[1890]: /opt/go/src/runtime/panic.go:464 +0x3e6
May 02 11:11:39 <host> nomad[1890]: github.com/hashicorp/nomad/client.(*AllocRunner).Run(0xc82018a1a0)
May 02 11:11:39 <host> nomad[1890]: /opt/gopath/src/github.com/hashicorp/nomad/client/alloc_runner.go:354 +0x9b
May 02 11:11:39 <host> nomad[1890]: created by github.com/hashicorp/nomad/client.(*Client).restoreState
May 02 11:11:39 <host> nomad[1890]: /opt/gopath/src/github.com/hashicorp/nomad/client/client.go:433 +0x8de
May 02 11:11:39 <host> systemd[1]: nomad.service: main process exited, code=exited, status=2/INVALIDARGUMENT
May 02 11:11:39 <host> systemd[1]: Unit nomad.service entered failed state.
May 02 11:11:39 <host> systemd[1]: nomad.service failed.
```

### Reproduction steps
This would be a tall order.

### Nomad Server logs (if appropriate)
see above

### Job file (if appropriate)
N/A

**Comment by dadgar** [comment]
Hey,

Thanks for reporting the bug. In the meantime for you to recover, you can delete the data under: /var/lib/nomad/ and restart the client.

**Comment by dvusboy** [comment]
That did it. Thanks. One issue is that the node now shows up twice when I do a `nomad node-status`. Obviously, after nuking stuff under `/var/lib/nomad`, the node now register with a new ID. Would this be just a cosmetic issue?

**Comment by dadgar** [comment]
@dvusboy Yes purely cosmetic, the old node-id will be GC'd

**Comment by dvusboy** [comment]
In that case, I don't think GC is working correctly. That no-long-exists client stays in the list reported by `nomad node-status` for a few days until I ran out of disk space and basically have to nuke the cluster and starts afresh (I posted this in the mailing list).

**Comment by dadgar** [comment]
@dvusboy It is slightly more complicated than a simple time cut off. If the Node that *should* be garbage collected had an allocation that is from a running batch job we don't GC the node because we need to keep that allocation around until the batch job is finished for proper accounting. And since the allocation is linked to the node, we don't GC the node.

**Comment by dadgar** [comment]
@dvusboy Do you by chance have the full stack trace?

**Comment by dvusboy** [comment]
I don't believe so. What's logged to `journald` is what I got. I don't see anything in `/var/log` for nomad.

**Comment by parasyte** [comment]
FWIW, 0.3.2

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x78 pc=0x5643f8]

goroutine 1 [running]:
panic(0x1124d20, 0xc82000e100)
    /opt/go/src/runtime/panic.go:464 +0x3e6
github.com/hashicorp/nomad/client/driver.GetTaskEnv(0xc82031b620, 0xc8200c0240, 0x0, 0xc820306f00, 0x0, 0x0, 0x0)
    /opt/gopath/src/github.com/hashicorp/nomad/client/driver/driver.go:145 +0xf8
github.com/hashicorp/nomad/client.(*TaskRunner).setTaskEnv(0xc82007b300, 0x0, 0x0)
    /opt/gopath/src/github.com/hashicorp/nomad/client/task_runner.go:203 +0x5a
github.com/hashicorp/nomad/client.(*TaskRunner).RestoreState(0xc82007b300, 0x0, 0x0)
    /opt/gopath/src/github.com/hashicorp/nomad/client/task_runner.go:140 +0x122
github.com/hashicorp/nomad/client.(*AllocRunner).RestoreState(0xc8203228f0, 0x0, 0x0)
    /opt/gopath/src/github.com/hashicorp/nomad/client/alloc_runner.go:129 +0x593
github.com/hashicorp/nomad/client.(*Client).restoreState(0xc82021d0e0, 0x0, 0x0)
    /opt/gopath/src/github.com/hashicorp/nomad/client/client.go:429 +0x53b
github.com/hashicorp/nomad/client.NewClient(0xc820234000, 0xc820234000, 0x0, 0x0)
    /opt/gopath/src/github.com/hashicorp/nomad/client/client.go:171 +0x8b9
github.com/hashicorp/nomad/command/agent.(*Agent).setupClient(0xc820214780, 0x0, 0x0)
    /opt/gopath/src/github.com/hashicorp/nomad/command/agent/agent.go:290 +0x173
github.com/hashicorp/nomad/command/agent.NewAgent(0xc8201ab400, 0x7f2fa894f650, 0xc8202170e0, 0xc8202146e0, 0x0, 0x0)
    /opt/gopath/src/github.com/hashicorp/nomad/command/agent/agent.go:55 +0x247
github.com/hashicorp/nomad/command/agent.(*Command).setupAgent(0xc8201a3220, 0xc8201ab400, 0x7f2fa894f650, 0xc8202170e0, 0x0, 0x0)
    /opt/gopath/src/github.com/hashicorp/nomad/command/agent/command.go:286 +0xc3
github.com/hashicorp/nomad/command/agent.(*Command).Run(0xc8201a3220, 0xc82000a0c0, 0x3, 0x3, 0x0)
    /opt/gopath/src/github.com/hashicorp/nomad/command/agent/command.go:389 +0x597
github.com/hashicorp/nomad/vendor/github.com/mitchellh/cli.(*CLI).Run(0xc820152480, 0xc820152480, 0x0, 0x0)
    /opt/gopath/src/github.com/hashicorp/nomad/vendor/github.com/mitchellh/cli/cli.go:153 +0x5ee
main.RunCustom(0xc82000a0b0, 0x4, 0x4, 0xc8201d12c0, 0xc82005a058)
    /opt/gopath/src/github.com/hashicorp/nomad/main.go:49 +0x4d5
main.Run(0xc82000a0b0, 0x4, 0x4, 0x7f2fa8985028)
    /opt/gopath/src/github.com/hashicorp/nomad/main.go:15 +0x4c
main.main()
    /opt/gopath/src/github.com/hashicorp/nomad/main.go:11 +0x60
```

**Comment by dadgar** [comment]
This has been fixed in 0.4.0
